server/service/src/programs/patient/search_central.rs

```diff
@@ -0,0 +1,129 @@
+use repository::RepositoryError;
+use reqwest::ClientBuilder;
+use url::Url;
+
+use crate::{
+    apis::patient_v4::{
+        NameStoreJoinParamsV4, NameStoreJoinV2, PatientApiV4, PatientParamsV4, PatientV4,
+    },
+    service_provider::{ServiceContext, ServiceProvider},
+    sync::settings::SyncSettings,
+};
+
+use super::PatientSearch;
+
+#[derive(Debug)]
+pub enum CentralPatientRequestError {
+    DatabaseError(RepositoryError),
+    InternalError(String),
+    ConnectionError(String),
+}
+
+pub async fn patient_search_central(
+    sync_settings: &SyncSettings,
+    params: PatientSearch,
+) -> Result<Vec<PatientV4>, CentralPatientRequestError> {
+    let central_server_url = Url::parse(&sync_settings.url).map_err(|err| {
+        CentralPatientRequestError::InternalError(format!(
+            "Failed to parse central server url: {}",
+            err
+        ))
+    })?;
+    let client = ClientBuilder::new()
+        .build()
+        .map_err(|err| CentralPatientRequestError::ConnectionError(format!("{:?}", err)))?;
+
+    let api = PatientApiV4::new(
+        client,
+        central_server_url.clone(),
+        &sync_settings.username,
+        &sync_settings.password_sha256,
+    );
+
+    let PatientSearch {
+        code,
+        code_2: _,
+        first_name,
+        last_name,
+        date_of_birth,
+        gender: _,
+    } = params;
+    let patients = api
+        .patient(PatientParamsV4 {
+            limit: None,
+            offset: None,
+            first_name,
+            last_name,
+            dob: date_of_birth,
+            policy_number: None,
+            barcode: None,
+            is_deleted: Some(false),
+            code,
+        })
+        .await
+        .map_err(|err| CentralPatientRequestError::ConnectionError(format!("{:?}", err)))?;
+    Ok(patients)
+}
+
+#[derive(Clone, Debug)]
+pub struct NameStoreJoin {
+    pub id: String,
+    pub name_id: String,
+    pub store_id: String,
+}
+
+/// Creates a name_store_join for the patient
+pub async fn link_patient_to_store(
+    service_provider: &ServiceProvider,
+    context: ServiceContext,
+    store_id: &str,
+    name_id: &str,
+) -> Result<(NameStoreJoin, ServiceContext), CentralPatientRequestError> {
```

Comment:
Don't need to return ServiceContext

Reply:
Yes did it on purpose to show that we just "borrowed" it by sending it back end forth and that in this way the ServiceContext can be used by another service... However, this is kind of trivial and can be added if needed. I removed it.